File: 02 백준/14 누적합/Python/Prob2167.py

"""
title  : 2167. 2차원 배열의 합 (Silver 5)
time   : 280ms        196ms
memory : 37028KB      112532KB
"""

import sys

def solution():
    
    input = sys.stdin.readline
    N, M = map(int, input().split())
    array = [list(map(int, input().split())) for _ in range(N)]

    prefix_sum = [[0] * M for _ in range(N)]

    for r in range(N):
        prefix_sum[r][0] = array[r][0]
        for c in range(1, M):
            prefix_sum[r][c] = prefix_sum[r][c-1] + array[r][c]

    K = int(input())
    
    for _ in range(K):
        i, j, x, y = map(int, input().split())
        print(search(prefix_sum, i, j, x, y))


def search(prefix_sum, i, j, x, y):

    total_sum = 0

    for idx in range(i-1, x):
        total_sum += (prefix_sum[idx][y-1] if j == 1 else prefix_sum[idx][y-1] - prefix_sum[idx][j-2])
    
    return total_sum


solution()

File: 02 백준/14 누적합/Java/src/Prob2167.java

/**
* title  : 2167. 2차원 배열의 합 (Silver 5)
* time   : ms
* memory : KB
*/

import java.io.*;
import java.util.*;

public class Prob2167 {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[][] board = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[][] answer = new int[N][M];
        
        answer[0][0] = board[0][0];

        for(int i = 1; i < N; i++) {
            answer[i][0] = board[i][0] + answer[i-1][0];
        }

        for(int j = 1; j < M; j++) {
            answer[0][j] = board[0][j] + answer[0][j-1];
        }

        for (int i = 1; i < N; i++) {
            for (int j = 1; j < M; j++) {
                answer[i][j] = board[i][j] + answer[i-1][j] + answer[i][j-1] - answer[i-1][j-1];
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                System.out.print(answer[i][j] + " ");
            }
            System.out.println();
        }

        int K = Integer.parseInt(br.readLine());
        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int r1 = Integer.parseInt(st.nextToken())-1;
            int c1 = Integer.parseInt(st.nextToken())-1;
            int r2 = Integer.parseInt(st.nextToken());
            int c2 = Integer.parseInt(st.nextToken()); 
            sb.append(answer[r2][c2] - answer[r1][c1]).append("\n");
        }

        System.out.println(sb);
    }
}